File: grpc/examples/python/submit-amend-pegged-order.py
#!/usr/bin/python3

"""
Script language: Python3

Talks to:
- Vega wallet (REST)
- Vega node (gRPC)

Apps/Libraries:
- Vega-API-client (https://pypi.org/project/Vega-API-client/)
"""

# Note: this file uses smart-tags in comments to section parts of the code to
# show them as snippets in our documentation. They are not necessary to be
# included when creating your own custom code.
#
# Example of smart-tags:
#  __something:
# some code here
# :something__
#

import base64
import helpers
import time
import requests

from google.protobuf.wrappers_pb2 import Int64Value

# Vega wallet interaction helper, see login.py for detail
from login import token, pubkey

# Load Vega network URLs, these are set using 'source examples-config'
# located in the root folder of the api repository
wallet_server_url = helpers.get_from_env("WALLETSERVER_URL")
node_url_grpc = helpers.get_from_env("NODE_URL_GRPC")

# __import_client:
import vegaapiclient as vac

# Vega gRPC clients for reading/writing data
data_client = vac.VegaTradingDataClient(node_url_grpc)
trading_client = vac.VegaTradingClient(node_url_grpc)
# :import_client__

#####################################################################################
#                               F I N D   M A R K E T                               #
#####################################################################################

# __get_market:
# Request the identifier for the market to place on
markets = data_client.Markets(vac.api.trading.MarketsRequest()).markets
marketID = markets[0].id
# :get_market__

assert marketID != ""
print(f"Market found: {marketID}")

#####################################################################################
#                          B L O C K C H A I N   T I M E                            #
#####################################################################################

# __get_expiry_time:
# Request the current blockchain time, calculate an expiry time
blockchain_time = data_client.GetVegaTime(vac.api.trading.GetVegaTimeRequest()).timestamp
expiresAt = int(blockchain_time + 120 * 1e9)  # expire in 2 minutes
# :get_expiry_time__

assert blockchain_time > 0
print(f"Blockchain time: {blockchain_time}")

#####################################################################################
#                      S U B M I T   P E G G E D   O R D E R                        #
#####################################################################################

# __prepare_submit_pegged_order:
# Prepare a submit order message with a pegged BUY order
order = vac.api.trading.PrepareSubmitOrderRequest(
    submission=vac.commands.v1.commands.OrderSubmission(
        market_id=marketID,
        side=vac.vega.Side.SIDE_BUY,
        size=50,
        expires_at=expiresAt,
        time_in_force=vac.vega.Order.TimeInForce.TIME_IN_FORCE_GTT,
        type=vac.vega.Order.Type.TYPE_LIMIT,
        pegged_order=vac.vega.PeggedOrder(
            offset=-5,
            reference=vac.vega.PEGGED_REFERENCE_MID
        )
    )
)
prepared_order = trading_client.PrepareSubmitOrder(order)
# :prepare_submit_pegged_order__

order_ref = prepared_order.submit_id
print(f"Prepared pegged order, ref: {order_ref}")

# __sign_tx_pegged_order:
# Sign the prepared pegged order transaction
# Note: Setting propagate to true will submit to a Vega node
blob_base64 = base64.b64encode(prepared_order.blob).decode("ascii")
req = {"tx": blob_base64, "pubKey": pubkey, "propagate": True}
url = f"{wallet_server_url}/api/v1/messages"
headers = {"Authorization": f"Bearer {token}"}
response = requests.post(url, headers=headers, json=req)
helpers.check_response(response)
signedTx = response.json()["signedTx"]
# :sign_tx_pegged_order__

print("Signed pegged order and sent to Vega")

# Wait for order submission to be included in a block
print("Waiting for blockchain...")
time.sleep(4)
order_ref_request = vac.api.trading.OrderByReferenceRequest(reference=order_ref)
response = data_client.OrderByReference(order_ref_request)
orderID = response.order.id
orderStatus = helpers.enum_to_str(vac.vega.Order.Status, response.order.status)
createVersion = response.order.version
orderReason = response.order.reason

print(f"\nPegged order processed, ID: {orderID}, Status: {orderStatus}, Version: {createVersion}")

if orderStatus == "STATUS_REJECTED":
    print(f"Rejection reason: {orderReason}")
else:
    print(f"Pegged at:\n{response.order.pegged_order}")

#####################################################################################
#                        A M E N D   P E G G E D   O R D E R                        #
#####################################################################################

# __prepare_amend_pegged_order:
# Prepare the amend pegged order message
amend = vac.commands.v1.commands.OrderAmendment(
    market_id=marketID,
    order_id=orderID,
    size_delta=-25,
    time_in_force=vac.vega.Order.TimeInForce.TIME_IN_FORCE_GTC,
    pegged_reference=vac.vega.PEGGED_REFERENCE_BEST_BID,
    pegged_offset=Int64Value(value=-100)
)
order = vac.api.trading.PrepareAmendOrderRequest(amendment=amend)
prepared_order = trading_client.PrepareAmendOrder(order)
blob_base64 = base64.b64encode(prepared_order.blob).decode("ascii")
# :prepare_amend_pegged_order__

print(f"Amendment prepared for order ID: {orderID}")

# __sign_tx_pegged_amend:
# Sign the prepared pegged order transaction for amendment
# Note: Setting propagate to true will also submit to a Vega node
req = {"tx": blob_base64, "pubKey": pubkey, "propagate": True}
url = f"{wallet_server_url}/api/v1/messages"
headers = {"Authorization": f"Bearer {token}"}
response = requests.post(url, headers=headers, json=req)
helpers.check_response(response)
signedTx = response.json()["signedTx"]
# :sign_tx_pegged_amend__

print("Signed pegged order amendment and sent to Vega")

# Wait for amendment to be included in a block
print("Waiting for blockchain...")
time.sleep(4)
order_id_request = vac.api.trading.OrderByIDRequest(order_id=orderID)
response = data_client.OrderByID(order_id_request)
orderID = response.order.id
orderPrice = response.order.status
orderSize = response.order.size
orderTif = helpers.enum_to_str(vac.vega.Order.TimeInForce, response.order.time_in_force)
orderStatus = helpers.enum_to_str(vac.vega.Order.Status, response.order.status)
orderVersion = response.order.version
orderReason = response.order.reason

print("Amended pegged order:")
print(f"ID: {orderID}, Status: {orderStatus}, "
      f"Size(Old): 50, Size(New): {orderSize}, "
      f"TimeInForce(Old): TIME_IN_FORCE_GTT, TimeInForce(New): {orderTif}, "
      f"Version(Old): {createVersion}, Version(new): {orderVersion}")

if orderStatus == "STATUS_REJECTED":
    print(f"Rejection reason: {orderReason}")
else:
    print(f"Pegged at:\n{response.order.pegged_order}")

# Completed.

File: grpc/examples/go/submit-amend-pegged-order.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"time"

	uuid "github.com/satori/go.uuid"
	"github.com/vegaprotocol/api/grpc/clients/go/generated/code.vegaprotocol.io/vega/proto"
	"github.com/vegaprotocol/api/grpc/clients/go/generated/code.vegaprotocol.io/vega/proto/api"
	v1 "github.com/vegaprotocol/api/grpc/clients/go/generated/code.vegaprotocol.io/vega/proto/commands/v1"
	"github.com/vegaprotocol/api/grpc/examples/go/helpers"

	"github.com/golang/protobuf/ptypes/wrappers"
	"golang.org/x/net/context"
	"google.golang.org/grpc"
	
)

func main() {
	// Helpers include logic to authenticate with a Vega wallet service
	// including token storage/public key operations/signing of commands
	// visit ./helpers/wallet.go for more detail
	pubkey := helpers.GetPubKey()

	nodeURLGrpc := helpers.GetFromEnv("NODE_URL_GRPC")
	conn, err := grpc.Dial(nodeURLGrpc, grpc.WithInsecure())
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	dataClient := api.NewTradingDataServiceClient(conn)

	// __get_market:
	// Request the identifier for the market to place on
	marketRequest := api.MarketsRequest{}
	markets, err := dataClient.Markets(context.Background(), &marketRequest)
	if err != nil {
		panic(err)
	}
	marketId := markets.Markets[0].Id
	// :get_market__

	// __get_expiry_time:
	// Request the current blockchain time, calculate an expiry time
	request := api.GetVegaTimeRequest{}
	vegaTime, err := dataClient.GetVegaTime(context.Background(), &request)

	expireAt := vegaTime.Timestamp + (120 * 1e9)
	// :get_expiry_time__

	fmt.Printf("Blockchain time: %d\n", vegaTime.Timestamp)
	fmt.Printf("Order expiration time: %d\n", expireAt)

	// __prepare_submit_pegged_order:
	// Compose the submit order message with a pegged BUY order
	orderRef := fmt.Sprintf("%s-%s", pubkey, uuid.NewV4())
	peggedOrder := proto.PeggedOrder{
		Offset:    -5,
		Reference: proto.PeggedReference_PEGGED_REFERENCE_MID,
	}
	orderSubmission := &v1.OrderSubmission{
		Size:        1,
		Price:       100000,
		MarketId:    marketId,
		Side:        proto.Side_SIDE_BUY,
		TimeInForce: proto.Order_TIME_IN_FORCE_GTT,
		Type:        proto.Order_TYPE_LIMIT,
		ExpiresAt:   expireAt,
		PeggedOrder: &peggedOrder,
		Reference:   orderRef,
	}
	// :prepare_submit_pegged_order__

	log.Printf("Order submission: %v\n", orderSubmission)

	// __sign_tx_order:
	reqCmd, _ := json.Marshal(orderSubmission)
	reqStr := fmt.Sprintf(
		"{ \"pub_key\": \"%s\", \"propagate\": true, \"order_submission\": %s }",
		pubkey, reqCmd)

	_, err = helpers.SignTransactionCommand(reqStr)
	if err != nil {
		panic(err)
	}
	// :sign_tx_order__

	fmt.Printf("Signed pegged order and sent to Vega\n")

	// Wait for order submission to be included in a block
	fmt.Printf("Waiting for blockchain...\n")
	time.Sleep(4 * time.Second)
	orderByRef := api.OrderByReferenceRequest{Reference: orderRef}

	orderByRefResp, err := dataClient.OrderByReference(context.Background(), &orderByRef)
	if err != nil {
		panic(err)
	}

	orderID := orderByRefResp.Order.Id
	orderStatus := orderByRefResp.Order.Status.String()
	fmt.Printf("Pegged order processed. ID: %s, Status: %s\n", orderID, orderStatus)

	// __prepare_amend_pegged_order:
	// Compose the amend pegged order message
	var peggedOffset wrappers.Int64Value
	peggedOffset.Value = -100

	orderAmendment := &v1.OrderAmendment{
		MarketId:        marketId,
		OrderId:         orderID,
		SizeDelta:       -25,
		TimeInForce:     proto.Order_TIME_IN_FORCE_GTC,
		PeggedReference: proto.PeggedReference_PEGGED_REFERENCE_BEST_BID,
		PeggedOffset:    &peggedOffset,
	}
	// :prepare_amend_pegged_order__
	log.Printf("Order amendment: %v\n", orderAmendment)

	// __sign_tx_amend:
	reqCmd, _ = json.Marshal(orderAmendment)
	reqStr = fmt.Sprintf(
		"{ \"pub_key\": \"%s\", \"propagate\": true, \"order_amendment\": %s }",
		pubkey, reqCmd)

	_, err = helpers.SignTransactionCommand(reqStr)
	if err != nil {
		panic(err)
	}
	// :sign_tx_amend__

	fmt.Printf("Signed pegged order amendment and sent to Vega\n")

	// Wait for amendment to be included in a block
	fmt.Printf("Waiting for blockchain...\n")
	time.Sleep(4 * time.Second)
	orderByRef = api.OrderByReferenceRequest{Reference: orderRef}
	orderByRefResp, err = dataClient.OrderByReference(context.Background(), &orderByRef)
	if err != nil {
		panic(err)
	}

	orderID = orderByRefResp.Order.Id
	orderStatus = orderByRefResp.Order.Status.String()
	oderSize := orderByRefResp.Order.Size
	orderTif := orderByRefResp.Order.TimeInForce.String()
	peggedOrderRef := orderByRefResp.Order.PeggedOrder

	fmt.Printf("Amended pegged order:\n")
	fmt.Printf("Pegged order processed. ID: %s, Status: %s\n", orderID, orderStatus)
	fmt.Printf("Size(Old): 50, Size(New): %d,\n", oderSize)
	fmt.Printf("TimeInForce(Old): TIME_IN_FORCE_GTT, TimeInForce(New): %s,\n", orderTif)
	fmt.Printf("Pegged at:\n%s\n", peggedOrderRef)
}
